GCP API Activation error on first deployment

Failed to deploy a project to GCP on Github action when doing an Defang up. 

```shell
Run set -o pipefail
  set -o pipefail
  params=()
  for filename in $COMPOSE_FILES; do
      params+=("-f")
      params+=("$filename")
  done
  if [[ -n "" ]]; then
    params+=("--mode=")
  fi
  
  case "false" in
    y|Y|yes|Yes|YES|true|True|TRUE|on|On|ON|1) params+=("--verbose") ;;
  esac
  
  echo defang $COMMAND "${params[@]}"
  output=$(defang $COMMAND "${params[@]}" | tee /dev/stderr)
  echo "stdout<<DEFANG_EOF" >> "$GITHUB_OUTPUT"
  echo "$output" >> "$GITHUB_OUTPUT"
  echo "DEFANG_EOF" >> "$GITHUB_OUTPUT"
  shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
  env:
    CLOUDSDK_AUTH_CREDENTIAL_FILE_OVERRIDE: /home/runner/work/nextjs/nextjs/gha-creds-7a35bec4f2fba4fc.json
    GOOGLE_APPLICATION_CREDENTIALS: /home/runner/work/nextjs/nextjs/gha-creds-7a35bec4f2fba4fc.json
    GOOGLE_GHA_CREDS_PATH: /home/runner/work/nextjs/nextjs/gha-creds-7a35bec4f2fba4fc.json
    DEFANG_DEBUG: 1
    DEFANG_STACK: production
    COMPOSE_PROJECT_NAME: nextjs
    COMMAND: up
    COMPOSE_FILES: 
##[debug]/usr/bin/bash --noprofile --norc -e -o pipefail /home/runner/work/_temp/47f5b297-9846-42eb-b1f2-4b6f59dcbb51.sh
defang up
 - Using tenant "personal tenant" for cluster "fabric-prod1.defang.dev:443"
 - Reading access token from file /home/runner/.local/state/defang/fabric-prod1.defang.dev
 - fabricClient slb4qishkogh /io.defang.v1.FabricController/WhoAmI null
 - fabricClient n947rm5vpfwp /io.defang.v1.FabricController/GetVersion null
 - Fabric: v0.6.0-1603-g1e80e294 CLI: 3.5.3 CLI-Min: v3.1.0
 - fabricClient eoqd8fb0sebd /io.defang.v1.FabricController/CheckToS null
 - tracking event "UP INVOKED": [{args []} {err <nil>} {non-interactive true} {provider auto} {CalledAs  up} {version 3.5.3}]
 - fabricClient juej71ulbqv8 /io.defang.v1.FabricController/Track {"anon_id":"3ac253f9-3b65-42d8-b0c2-52653acd4d3d","event":"UP INVOKED","properties":{"CalledAs":" up","args":"[]","err":"\u003cnil\u003e","non-interactive":"true","provider":"auto","version":"3.5.3"},"os":"linux","arch":"amd64"}
name: nextjs
services:
    app:
        build:
            context: /home/runner/work/nextjs/nextjs/app
            dockerfile: Dockerfile
        deploy:
            resources:
                reservations:
                    cpus: 0.25
                    memory: "268435456"
        healthcheck:
            test:
                - CMD-SHELL
                - wget -q --spider http://localhost:3000/ || exit 1
        networks:
            default: null
        ports:
            - mode: ingress
              target: 3000
              published: "3000"
              protocol: tcp
        restart: unless-stopped
networks:
    default:
        name: nextjs_default

 - Creating gcp provider
 - checking if application default credentials are available and has permission, GOOGLE_APPLICATION_CREDENTIALS="/home/runner/work/nextjs/nextjs/gha-creds-7a35bec4f2fba4fc.json"...
 * Using the "production" stack on gcp from DEFANG_STACK environment variable and local stack file
 - Enabling services: [cloudresourcemanager.googleapis.com]
 - found valid application default credentials with required permissions
name: nextjs
services:
    app:
        build:
            context: /home/runner/work/nextjs/nextjs/app
            dockerfile: Dockerfile
        deploy:
            resources:
                reservations:
                    cpus: 0.25
                    memory: "268435456"
        healthcheck:
            test:
                - CMD-SHELL
                - wget -q --spider http://localhost:3000/ || exit 1
        networks:
            default: null
        ports:
            - mode: ingress
              target: 3000
              published: "3000"
              protocol: tcp
        restart: unless-stopped
networks:
    default:
        name: nextjs_default

 - Getting services from bucket: defang-cd-0hjyntgkvnik projects/nextjs/production/project.pb defang-upload@cloudbuildtest-468719.iam.gserviceaccount.com
 - Failed to get project bucket object from bucket "defang-cd-0hjyntgkvnik" at path "projects/nextjs/production/project.pb" with service account "defang-upload@cloudbuildtest-468719.iam.gserviceaccount.com": failed to get bucket object ("projects/nextjs/production/project.pb") reader: Get "https://storage.googleapis.com/defang-cd-0hjyntgkvnik/projects%2Fnextjs%2Fproduction%2Fproject.pb": impersonate: status code 403: {
  "error": {
    "code": 403,
    "message": "Permission 'iam.serviceAccounts.getAccessToken' denied on resource (or it may not exist).",
    "status": "PERMISSION_DENIED",
    "details": [
      {
        "@type": "type.googleapis.com/google.rpc.ErrorInfo",
        "reason": "IAM_PERMISSION_DENIED",
        "domain": "iam.googleapis.com",
        "metadata": {
          "permission": "iam.serviceAccounts.getAccessToken"
        }
      }
    ]
  }
}
Error: failed to get bucket object ("projects/nextjs/production/project.pb") reader: Get "https://storage.googleapis.com/defang-cd-0hjyntgkvnik/projects%2Fnextjs%2Fproduction%2Fproject.pb": impersonate: status code 403: {
  "error": {
    "code": 403,
    "message": "Permission 'iam.serviceAccounts.getAccessToken' denied on resource (or it may not exist).",
    "status": "PERMISSION_DENIED",
    "details": [
      {
        "@type": "type.googleapis.com/google.rpc.ErrorInfo",
        "reason": "IAM_PERMISSION_DENIED",
        "domain": "iam.googleapis.com",
        "metadata": {
          "permission": "iam.serviceAccounts.getAccessToken"
        }
      }
    ]
  }
}

 - tracking event "CLI Error": [{err failed to get bucket object ("projects/nextjs/production/project.pb") reader: Get "https://storage.googleapis.com/defang-cd-0hjyntgkvnik/projects%2Fnextjs%2Fproduction%2Fproject.pb": impersonate: status code 403: {
  "error": {
    "code": 403,
    "message": "Permission 'iam.serviceAccounts.getAccessToken' denied on resource (or it may not exist).",
    "status": "PERMISSION_DENIED",
    "details": [
      {
        "@type": "type.googleapis.com/google.rpc.ErrorInfo",
        "reason": "IAM_PERMISSION_DENIED",
        "domain": "iam.googleapis.com",
        "metadata": {
          "permission": "iam.serviceAccounts.getAccessToken"
        }
      }
    ]
  }
}
}]
 - fabricClient j5fzrrnmmuyk /io.defang.v1.FabricController/Track {"anon_id":"3ac253f9-3b65-42d8-b0c2-52653acd4d3d","event":"CLI Error","properties":{"err":"failed to get bucket object (\"projects/nextjs/production/project.pb\") reader: Get \"[https://storage.googleapis.com/defang-cd-0hjyntgkvnik/projects%2Fnextjs%2Fproduction%2Fproject.pb\](https://storage.googleapis.com/defang-cd-0hjyntgkvnik/projects%2Fnextjs%2Fproduction%2Fproject.pb/)": impersonate: status code 403: {\n  \"error\": {\n    \"code\": 403,\n    \"message\": \"Permission 'iam.serviceAccounts.getAccessToken' denied on resource (or it may not exist).\",\n    \"status\": \"PERMISSION_DENIED\",\n    \"details\": [\n      {\n        \"@type\": \"type.googleapis.com/google.rpc.ErrorInfo\",\n        \"reason\": \"IAM_PERMISSION_DENIED\",\n        \"domain\": \"iam.googleapis.com\",\n        \"metadata\": {\n          \"permission\": \"iam.serviceAccounts.getAccessToken\"\n        }\n      }\n    ]\n  }\n}\n"},"os":"linux","arch":"amd64"}
Error: Process completed with exit code 2.
##[debug]Finished: run
```

Another example:
```
noumanusman@Muhammads-MacBook-Air defang-sample % cd nextjs-cv
noumanusman@Muhammads-MacBook-Air nextjs-cv % defang compose up
 ! Please log in to continue.
Please visit the following URL to log in: (Right click the URL or press ENTER to open browser)
  https://auth.defang.io/cli/-I_pgYRkPF9bG_QfJDq_Xw/TjxnWV4-fTwqPNOMiB7vwG2pZ-EtTrbNnvaoH1cfpXA
                                                                                Where do you want to deploy?
? provider Google Cloud Platform
Which region do you want to deploy to?
? region us-central1
What do you want to call this stack?:
? stack_name dev
What is your GCP Project ID?:
? gcp_project_id defang-test-493108
 
Using the "dev" stack on gcp from interactive selection
Please visit the following URL to log in to Google Cloud Platform: (Right click the URL or press ENTER to open browser)
  https://auth.defang.io/gcp/Hmy1OUImoxyZ1GUKMJ2IbK8THaMf4Y9pp8sRxxB711w=

Error: failed to get bucket object ("projects/nextjs-cv/dev/project.pb") reader: Get "https://storage.googleapis.com/defang-cd-sya02zxg1tz4/projects%2Fnextjs-cv%2Fdev%2Fproject.pb": impersonate: status code 403: {
  "error": {
    "code": 403,
    "message": "Permission 'iam.serviceAccounts.getAccessToken' denied on resource (or it may not exist).",
    "status": "PERMISSION_DENIED",
    "details": [
      {
        "@type": "type.googleapis.com/google.rpc.ErrorInfo",
        "reason": "IAM_PERMISSION_DENIED",
        "domain": "iam.googleapis.com",
        "metadata": {
          "permission": "iam.serviceAccounts.getAccessToken"
        }
      }
    ]
  }
}
```

It seem even after a day both me and the user cannot get past this error. Maybe it only occurs on OIDC: https://github.com/KevyVo/nextjs/actions/runs/24491265471

--- a/src/bin/install
+++ b/src/bin/install
@@ -286,9 +286,11 @@ _install_completion_script() {
     esac
 }
 
-_install_completion_script "$SHELL" || true
-CURRENT_SHELL=${0#-}
-[ "$CURRENT_SHELL" != "$SHELL" ] && _install_completion_script "$CURRENT_SHELL" || true
+if [ "$CI" != "1" ]; then
+    _install_completion_script "$SHELL" || true
+    CURRENT_SHELL=${0#-}
+    [ "$CURRENT_SHELL" != "$SHELL" ] && _install_completion_script "$CURRENT_SHELL" || true
+fi
 
 # Cleanup: Remove the temporary directory
 echo "Cleaning up..."

--- a/src/pkg/cli/client/byoc/aws/byoc_test.go
+++ b/src/pkg/cli/client/byoc/aws/byoc_test.go
@@ -94,18 +94,6 @@ func TestDomainMultipleProjectSupport(t *testing.T) {
 	}
 }
 
-type FakeLoader struct {
-	ProjectName string
-}
-
-func (f FakeLoader) LoadProject(ctx context.Context) (*composeTypes.Project, error) {
-	return &composeTypes.Project{Name: f.ProjectName}, nil
-}
-
-func (f FakeLoader) LoadProjectName(ctx context.Context) (string, bool, error) {
-	return f.ProjectName, false, nil
-}
-
 //go:embed testdata/*.json
 var testDir embed.FS
 

--- a/src/pkg/cli/client/byoc/gcp/byoc.go
+++ b/src/pkg/cli/client/byoc/gcp/byoc.go
@@ -282,7 +282,6 @@ func (b *ByocGcp) SetUpCD(ctx context.Context, force bool) error {
 		}
 	}
 
-	// 5. Setup Cloud Run Job
 	term.Debugf("Using CD image: %q", b.CDImage)
 
 	b.SetupDone = true
@@ -847,10 +846,12 @@ func (b *ByocGcp) GetProjectUpdate(ctx context.Context, projectName string) (*de
 	pbBytes, err := b.driver.GetBucketObjectWithServiceAccount(ctx, bucketName, path, uploadSA)
 	if err != nil {
 		term.Debugf("Failed to get project bucket object from bucket %q at path %q with service account %q: %v", bucketName, path, uploadSA, err)
-		if errors.Is(err, gcp.ErrObjectNotExist) {
-			return nil, client.ErrNotExist // no services yet
+		// Handle the case where the object does not exist, or where we do not have permission to view the object, ie.
+		// "Permission 'iam.serviceAccounts.getAccessToken' denied on resource (or it may not exist)."  #2051
+		if errors.Is(err, gcp.ErrObjectNotExist) || strings.Contains(err.Error(), "(or it may not exist)") {
+			return nil, client.ErrNotExist // first deployment, no services yet
 		}
-		return nil, err
+		return nil, annotateGcpError(err)
 	}
 
 	var projUpdate defangv1.ProjectUpdate

--- a/src/pkg/cli/client/byoc/gcp/byoc_test.go
+++ b/src/pkg/cli/client/byoc/gcp/byoc_test.go
@@ -17,6 +17,7 @@ import (
 	defangv1 "github.com/DefangLabs/defang/src/protos/io/defang/v1"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
+	"google.golang.org/api/googleapi"
 	"google.golang.org/protobuf/proto"
 	"google.golang.org/protobuf/types/known/timestamppb"
 )
@@ -364,6 +365,21 @@ func TestGetServices(t *testing.T) {
 		}
 		res, err := b.GetServices(t.Context(), &defangv1.GetServicesRequest{
 			Project: "project2",
+		})
+		require.NoError(t, err)
+		assert.Empty(t, res.Services)
+	})
+
+	t.Run("first deployment 403 = no services", func(t *testing.T) {
+		b.driver = &mockGcpDriver{
+			bucketName: "bucket-a",
+			getBucketObjectWithServiceAccountError: &googleapi.Error{
+				Code:    403,
+				Message: "Permission 'iam.serviceAccounts.getAccessToken' denied on resource (or it may not exist).",
+			},
+		}
+		res, err := b.GetServices(t.Context(), &defangv1.GetServicesRequest{
+			Project: "project1",
 		})
 		require.NoError(t, err)
 		assert.Empty(t, res.Services)

--- a/src/pkg/cli/client/byoc/gcp/errors.go
+++ b/src/pkg/cli/client/byoc/gcp/errors.go
@@ -3,7 +3,6 @@ package gcp
 import (
 	"errors"
 	"fmt"
-	"strings"
 
 	"github.com/DefangLabs/defang/src/pkg/http"
 	"google.golang.org/api/googleapi"
@@ -40,21 +39,4 @@ func annotateGcpError(err error) error {
 		return briefErr
 	}
 	return err
-}
-
-// Used to get nested values from the detail of a googleapi.Error
-func GetGoogleAPIErrorDetail(detail any, path string) string {
-	if path == "" {
-		value, ok := detail.(string)
-		if ok {
-			return value
-		}
-		return ""
-	}
-	dm, ok := detail.(map[string]any)
-	if !ok {
-		return ""
-	}
-	key, rest, _ := strings.Cut(path, ".")
-	return GetGoogleAPIErrorDetail(dm[key], rest)
 }

--- a/src/pkg/cli/client/byoc/gcp/errors_test.go
+++ b/src/pkg/cli/client/byoc/gcp/errors_test.go
@@ -65,7 +65,7 @@ func Test_BadGCPprojectnameErrorWrap(t *testing.T) {
 	wrappedErr := annotateGcpError(gcpErr)
 	assert.Equal(t, `double check the GCP project ID and make sure your Application Default Credentials have permission to access the project: `+gcpErr.Message, wrappedErr.Error())
 
-	// Test the error is wrapper correctly
+	// Test the error is wrapped correctly
 	var unwrappedErr *googleapi.Error
 	assert.True(t, errors.As(wrappedErr, &unwrappedErr))
 	assert.Equal(t, gcpErr, unwrappedErr)

